example-raft-kv/src/network/api.rs

```diff
@@ -20,9 +27,56 @@ use crate::store::ExampleRequest;
  */
 #[post("/write")]
 pub async fn write(app: Data<ExampleApp>, req: Json<ExampleRequest>) -> actix_web::Result<impl Responder> {
-    let request = ClientWriteRequest::new(EntryPayload::Normal(req.0));
-    let response = app.raft.client_write(request).await;
-    Ok(Json(response))
+
+    // Redirect this call to the leader, this is required because a `Candidate`, `Follower` or `Learner` is not allowed to write data to
+    // other nodes. The payload has to wrapped around the type fo expect payload that raft uses. In this case it is real payload and
+    // `EntryPayload::Normal(D)` will be used.
+    let request: ClientWriteRequest<ExampleRequest> = ClientWriteRequest::new(EntryPayload::Normal(req.0.clone()));
+    
+    // Attempt to write the data to the raft and dispatch to the nodes.
+    let response: Result<ClientWriteResponse<ExampleResponse>, ClientWriteError> = app.raft.client_write(request).await;
+
+    match response {
+        Ok(_) => Ok(HttpResponse::Ok().finish()),
+        Err(error) => {
+            match error {
+                ClientWriteError::ForwardToLeader(foward_to_leader) => 
+                    redirect_to_leader(&app, &req, "/write", foward_to_leader).await,
+                _ => Ok(HttpResponse::BadRequest().body("Failed write number"))
+            }
+        }
+    }
+}
+
+async fn redirect_to_leader<T: Serialize>(
+  app: &Data<ExampleApp>,
+  req: &Json<T>,
+  route: &str,
+  foward_to_leader: ForwardToLeader,
+) -> actix_web::Result<HttpResponse> {
+
+  // In a real world application you should not unwrap this, instead return a error related to the missing leader id.
+  let leader_id: NodeId = foward_to_leader.leader_id.unwrap();
```

Comment:
It is possible `foward_to_leader` is None because it may have not yet received an `append-entries` RPC.
Only when a node received `append-entries` RPC, the node can be sure there is an established leader.

Reply:
Shouldn't this be a fatal error on the node?

Reply:
No. It is a normal condition.  :(
Especially in a partitioned network, a node may stay in a Candidate state for a while. During this period, the known leader on this node is always `None`.

Reply:
Maybe instead of `Option`, use a `Result<NodeId, NodeIdError>` and handle these cases. Just a idea.

Reply:
To me, it's a little bit weird to have a `Result` in an `Error`. 🤔